cilium/cilium
# Spiffe Integration

### What is this?
This PR contains the Cilium-SPIFFE integration that has been proposed in [this](https://docs.google.com/document/d/1QB3JIomGmEqg34sCQCmn9A-RqrCwBJMQgj7hoFkwTE0/edit#heading=h.lzba1zbo2xvl) design document. This PR covers the "Identity Generation Hardening" part of such proposal. We already have some prototypes of the "Upgrading connections to mTLS" part but we prefer start discussing only the former one to keep it more scoped.

### How to test?
If you want to test it by yourself you can follow these steps. Otherwise you can watch [this](https://asciinema.org/a/421142) recorded demo.

0. Download, compile and install cilium-agent based on this PR
1. Add the `--enable-spiffe` flag to cilium-agent
2. Deploy Spire. I prepared a yaml manifest with all components needed to deploy it. It's based on the [spire k8s quickstart tutorial](https://github.com/spiffe/spire-tutorials/tree/master/k8s/quickstart) but uses a custom spire-agent image that implements a POC of the spire privileged API.
```
$ kubectl apply -f https://gist.githubusercontent.com/rscampos/1e1e537c5183ccc32f2f64a949f37f41/raw/c5543cd286be7dda5d3940a09129378465de0dac/spire.yaml
```
3. Check that spire pods are running
```
$ kubectl get pods -n spire
NAME                READY   STATUS    RESTARTS   AGE
spire-agent-rqsp9   1/1     Running   0          14s
spire-server-0      1/1     Running   0          14s
```
4. Create some registration entries. [This](https://gist.githubusercontent.com/rscampos/1e1e537c5183ccc32f2f64a949f37f41/raw/c5543cd286be7dda5d3940a09129378465de0dac/create_registration_entries.sh) creates some registration entries for testing.
```
$ ./create_registration_entries.sh
```
5. Create some pods
```
$ kubectl run podubuntu --image=ubuntu:latest -- sleep 1000000
$ kubectl run podpraqma --image=praqma/network-multitool:latest -- sleep 100000
```
6. Restrict all egress traffic
```
$ kubectl apply -f https://gist.githubusercontent.com/rscampos/1e1e537c5183ccc32f2f64a949f37f41/raw/c5543cd286be7dda5d3940a09129378465de0dac/deny-all-egress.yaml
```
7. Check that ping is failing
```
$ kubectl exec podpraqma -- ping <podubuntu IP>
```
8. Deploy image based policy that allows those two pods to talk to each other
```
$ kubectl apply -f https://gist.githubusercontent.com/rscampos/1e1e537c5183ccc32f2f64a949f37f41/raw/c5543cd286be7dda5d3940a09129378465de0dac/image-based.yaml
```
9. Ping now works
```
$ kubectl exec podpraqma -- ping <podubuntu IP>
PING 10.11.0.173 (10.11.0.173) 56(84) bytes of data.
64 bytes from 10.11.0.173: icmp_seq=1 ttl=63 time=0.124 ms
```

### TODO
- [x] Update code once Spire "privileged" PR is ready (https://github.com/spiffe/spire/issues/2246 & https://github.com/spiffe/spire-api-sdk/pull/8 & https://github.com/spiffe/spire/pull/2481)
- [x] Improve label validation (undo validation relax done for this POC)
- [x] Rebase once https://github.com/cilium/cilium/pull/15123 is merged
- [x] Rebase once https://github.com/cilium/cilium/pull/16585 is merged
- [x] Spiffe label format - use **spiffe** as source in Spiffe label - discussed here #16626 
- [ ] Tests (both unit and integration)
- [ ] How to deploy spire. Create helm templates?, leave it up to the user?
- [ ] How to initialize the spiffe watcher in a lazy way (avoid trying to connect to it when spire is not deployed yet)

ref #4016
old PR #16626

cc @nyrahul @jrajahalme @jrfastab @navarrothiago @mauriciovasquezbernal
